WIP: Use common storage config in jsonnet (#3257)
grafana/mimir: docs/sources/operators-guide/deploy-grafana-mimir/jsonnet/configuring-object-storage-backend.md

---
description: Learn how to configure the Grafana Mimir object storage backend when using Jsonnet.

> menuTitle: Configuring object storage

Reviewer: ```suggestion
menuTitle: Configure object storage
```

> description: Learn how to configure the Grafana Mimir object storage backend when using Jsonnet.
> menuTitle: Configuring object storage
> title: Configuring the Grafana Mimir object storage backend with Jsonnet
> weight: 20
> ---
> 
> # Configuring the Grafana Mimir object storage backend with Jsonnet

Reviewer: ```suggestion
description: Learn how to configure the Grafana Mimir object storage backend when using Jsonnet.
menuTitle: Configure object storage
title: Configure the Grafana Mimir object storage backend with Jsonnet
weight: 20
---

# Configure the Grafana Mimir object storage backend with Jsonnet
```

> The object storage backend can be configured for all Mimir components from a single place.

Reviewer: ```suggestion
You can configure the object storage backend for all Mimir components from a single place.
```

The minimum Jsonnet code required for this is:

```jsonnet
{
  _config+:: {
    storage_backend: 'gcs',
    blocks_storage_bucket_name: 'blocks-bucket',
  }
}
```

The `storage_backend` option must be one of either `azure`, `gcs`, or `s3`.

> For each one of those providers additional configuration options are required:

Reviewer: ```suggestion
For each one of those providers, additional configuration options are required:
```

- Amazon S3 (`s3`)

>   - `aws_region`
>   - `s3-secret-access-key
> 

Author: This shouldn't be here.
```suggestion
  - `aws_region`

```

- Azure (`azure`)

  - `storage_azure_account_name`
  - `storage_azure_account_key`

> **Note:** You need to manually provide the storage credentials for `s3` and `gcs` by using additional command line arguments as necessary.

> > For more information about different common storage configurations, see [Grafana Mimir configuration parameters: `common`]({{< relref "../../configure/reference-configuration-parameters/index.md#common" >}}) section.

Reviewer: ```suggestion
> For more information about different common storage configurations, see [Grafana Mimir configuration parameters: `common`]({{< relref "../../configure/reference-configuration-parameters/index.md#common" >}}).
```

>

> > For more information on how to provide GCS and S3 storage credentials see:

Reviewer: ```suggestion
> For more information about how to provide GCS and S3 storage credentials see the following resources:
```

  }
}
```

The `storage_backend` option must be one of either `azure`, `gcs`, or `s3`.
Additional configuration options are available for each one of these providers.

## Amazon S3 (`s3`) storage configuration options

Amazon S3 storage can be accessed without credentials when [using Amazon VPC](https://aws.amazon.com/premiumsupport/knowledge-center/s3-private-connection-no-authentication/).
In this case, `storage_s3_secret_access_key` and `storage_s3_access_key_id` are optional and can be left null, as in the following example:

```jsonnet
{
  _config+:: {
    storage_backend: 's3',
    blocks_storage_bucket_name: 'blocks-bucket',
    aws_region: 'af-south-1',
  }
}
```

If credentials are required, it is a good practice to keep them in secrets. In that case environment variable interpolation can be used:

```jsonnet
{
  _config+:: {
    storage_backend: 's3',
    storage_s3_access_key_id: '$(BLOCKS_STORAGE_S3_ACCESS_KEY_ID)',
    storage_s3_secret_access_key: '$(BLOCKS_STORAGE_S3_SECRET_ACCESS_KEY)',
    aws_region: 'af-south-1',
    blocks_storage_bucket_name: 'blocks-bucket',
  }
}
```

## Azure (`azure`) storage configuration options

> Azure storage client requires the `storagE_azurE_account_name` and `storage_azure_account_key` to be configured.

Reviewer 2: ```suggestion
Azure storage client requires the `storage_azure_account_name` and `storage_azure_account_key` to be configured.
```

Author: Oh God, how did this happen? How didn't I notice it?. Thank you.

Author: Fixed here as file was renamed: https://github.com/grafana/mimir/pull/3257/commits/538ad46c39d7e27d2f59495e09900ae152845645
